A29-xds-tls-security.md

```diff
@@ -0,0 +1,446 @@
+in this case is different from Envoy's due to fallback credentials. Envoy will just
+use plaintext (insecure) communication mode in this case. But with gRPC, the application
+needs to use InsecureCredentials as the fallback credentials to get the same result.
+
+The fallback credentials is *not* used in case of error situations e.g. when there is
+an error in using the xDS provided security configuration.
+
+Note that a user is not required to use Xds- Channel or Server Credentials even if they are
+using an xDS managed channel or server. Not using Xds Channel or Server Credentials results
+in not using xDS-provided TLS configuration.
+
+On the client side, an XdsChannelCredentials is used and on the server side an
+XdsServerCredentials is used as described in the example snippets below. In these examples,
+a plaintext or insecure credentials is used as the fallback credentials.
+
+
+#### C++ 
+
+Use of XdsChannelCredentials:
+
+```C++
+    std::shared_ptr<grpc::ChannelCredentials> credentials =
+        grpc::experimental::XdsCredentials(grpc::InsecureChannelCredentials());
+    std::shared_ptr<grpc::Channel> channel = grpc::CreateChannel(target, credentials);
+```
+
+Use of XdsServerCredentials:
+
+```C++
+   grpc::experimental::XdsServerBuilder builder;
+   builder.RegisterService(&service);
+   builder.AddListeningPort(listening_address,
+                                 grpc::experimental::XdsServerCredentials(
+                                     grpc::InsecureServerCredentials()));
+   builder.BuildAndStart()->Wait();
+```
+
+#### Java
+
+Use of XdsChannelCredentials:
+
+```Java
+    ChannelCredentials credentials = XdsChannelCredentials.create(InsecureChannelCredentials.create());
+    ManagedChannel channel = Grpc.newChannelBuilder(target, credentials).build();
+```
+
+Use of XdsServerCredentials:
+
+```Java
+    ServerCredentials credentials = XdsServerCredentials.create(InsecureServerCredentials.create());
+    Server server = XdsServerBuilder.forPort(port, credentials)
+      .addService(new HostnameGreeter(hostname))
+      .build()
+      .start();
+```
+
+#### Go
+
+Use of XdsChannelCredentials:
+
+```Go
+   import (
+           xdscreds "google.golang.org/grpc/credentials/xds"
+   )
+
+   if creds, err := xdscreds.NewClientCredentials(xdscreds.ClientOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+           log.Fatalf("Failed to create xDS client credentials: %v", err)
+   }
+   conn, err := grpc.Dial(*target, grpc.WithTransportCredentials(creds))
+```
+
+Use of XdsServerCredentials:
+
+```Go
+   import (
+           xdscreds "google.golang.org/grpc/credentials/xds"
+           "google.golang.org/grpc/xds"
+   )
+
+   if creds, err := xdscreds.NewServerCredentials(xdscreds.ServerOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+           log.Fatalf("Failed to create xDS server credentials: %v", err)
+   }
+   server := xds.NewGRPCServer(grpc.Creds(creds))
+```
+
+### xDS Protocol
```

Comment:
Sorry for being late to the party. But the sub-sections underneath and hard to read since they are structured as paragraphs. What do you think about using bullet points instead?

Reply:
Is that because you are trying to view it as a raw text file? How about using the markdown mode of the browser? Also what's the exact issue of "hard to read" due to paragraphs?

Reply:
I was viewing the markdown, the issue is not that. I feel that each paragraph is just a stream of words, with references to a lot of protos and fields within them. If we split them into points, and show relevant proto snippets with relevant fields, I feel it would make it easier to read.

Reply:
+1, had the same feeling. It's not a blocker, but bullet points would've been easier to follow.